suitcase/json_line_delimited/__init__.py

```diff
@@ -13,25 +13,57 @@
 # intended to be user-facing.
 
 
-def export(gen, directory, file_prefix, **kwargs):
+def export(gen, directory, file_prefix='{uid}', **kwargs):
     serializer = Serializer(directory, file_prefix, **kwargs)
     try:
         for item in gen:
             serializer(*item)
     finally:
         serializer.close()
-
     return serializer.artifacts
 
 
 class Serializer(event_model.DocumentRouter):
-    def __init__(self, directory, file_prefix, **kwargs):
+    def __init__(self, directory, file_prefix='{uid}', **kwargs):
+
+        self._output_file = None
+        self._file_prefix = file_prefix
+        self._templated_file_prefix = ''
+        self._kwargs = kwargs
+        self._start_found = False
+
         if isinstance(directory, (str, Path)):
-            self.manager = suitcase.utils.MultiFileManager(directory)
+            self._manager = suitcase.utils.MultiFileManager(directory)
         else:
-            self.manager = directory
-        self.artifacts = self.manager._artifacts  # Expose as public.
-        ...
+            self._manager = directory
+
+        self.artifacts = self._manager.artifacts
+
+    def _check_start(self, doc):
+        if self._start_found:
+            raise RuntimeError(
+                "The serializer in suitcase-json-line-delimited expects "
+                "documents from one run only. Two `start` documents where "
+                "sent to it")
+        else:
+            self._start_found = True
+            self._templated_file_prefix = self._file_prefix.format(**doc)
+
+    def _get_file(self):
+        filename = (f'{self._templated_file_prefix}.jsonl')
+        self._output_file = self._manager.open('stream_data', filename, 'xt')
+
+    def __call__(self, name, doc):
+        if name == 'start':
+            self._check_start(doc)
+            self._get_file()
+
+        self._output_file.write("%s\n" % json.dumps(doc))
+        return name, doc
+
+    def __del__(self):
```

Comment:
Thinking on this some more, I think this method is superfluous. The `__del__` method is called by Python when an object is explicitly deleted or garbage collected. In either case, once the `Serializer` instance is deleted, there ought to be zero remaining references to its `_output_file` attribute, so that will itself be garbage collected, and the garbage collector will call `__del__` on the file, which will close it. Therefore, we don't need to explicitly close it here.

Reply:
Updated